File: packages/php/src/Types/Metadata/DbfMetadata.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Types\Metadata;

/**
 * dBASE (DBF) file metadata.
 *
 * Contains record count, field count, and field definitions.
 */
readonly class DbfMetadata
{
    /**
     * @param DbfFieldInfo[] $fields
     */
    public function __construct(
        public int $recordCount = 0,
        public int $fieldCount = 0,
        public array $fields = [],
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        $rawRecordCount = $data['record_count'] ?? 0;
        $recordCount = is_int($rawRecordCount) ? $rawRecordCount : (is_numeric($rawRecordCount) ? (int) $rawRecordCount : 0);

        $rawFieldCount = $data['field_count'] ?? 0;
        $fieldCount = is_int($rawFieldCount) ? $rawFieldCount : (is_numeric($rawFieldCount) ? (int) $rawFieldCount : 0);

        /** @var DbfFieldInfo[] $fields */
        $fields = [];
        $rawFields = $data['fields'] ?? null;
        if (is_array($rawFields)) {
            foreach ($rawFields as $fieldData) {
                if (is_array($fieldData)) {
                    /** @var array<string, mixed> $typedFieldData */
                    $typedFieldData = $fieldData;
                    $fields[] = DbfFieldInfo::fromArray($typedFieldData);
                }
            }
        }

        return new self(
            recordCount: $recordCount,
            fieldCount: $fieldCount,
            fields: $fields,
        );
    }

    /**
     * @return array<string, mixed>
     */
    public function toArray(): array
    {
        $result = [
            'record_count' => $this->recordCount,
            'field_count' => $this->fieldCount,
        ];

        if ($this->fields !== []) {
            $result['fields'] = array_map(
                fn (DbfFieldInfo $f) => $f->toArray(),
                $this->fields,
            );
        }

        return $result;
    }
}

File: kreuzberg/packages/elixir/lib/kreuzberg/dbf_metadata.ex
defmodule Kreuzberg.DbfMetadata do
  @moduledoc """
  dBASE (DBF) file metadata.

  Matches the Rust `DbfMetadata` struct.

  ## Fields

    * `:record_count` - Number of records
    * `:field_count` - Number of fields
    * `:fields` - List of field definitions
  """

  @type t :: %__MODULE__{
          record_count: non_neg_integer(),
          field_count: non_neg_integer(),
          fields: list(Kreuzberg.DbfFieldInfo.t())
        }

  defstruct record_count: 0,
            field_count: 0,
            fields: []

  @spec from_map(map()) :: t()
  def from_map(data) when is_map(data) do
    fields =
      (data["fields"] || [])
      |> Enum.map(fn
        %Kreuzberg.DbfFieldInfo{} = f -> f
        map when is_map(map) -> Kreuzberg.DbfFieldInfo.from_map(map)
        _ -> nil
      end)
      |> Enum.reject(&is_nil/1)

    %__MODULE__{
      record_count: data["record_count"] || 0,
      field_count: data["field_count"] || 0,
      fields: fields
    }
  end

  @spec to_map(t()) :: map()
  def to_map(%__MODULE__{} = meta) do
    %{
      "record_count" => meta.record_count,
      "field_count" => meta.field_count,
      "fields" => Enum.map(meta.fields, &Kreuzberg.DbfFieldInfo.to_map/1)
    }
  end
end
